clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java

```diff
@@ -210,6 +220,38 @@ public void testMultipleServerMechanisms() throws Exception {
     }
 
     /**
+     * Tests that Kafka ApiVersionRequests are handled by the SASL server authenticator
+     * prior to SASL handshake flow and that subsequent authentication succeeds.
+     */
+    @Test
+    public void testUnauthenticatedApiVersionRequest() throws Exception {
+        SecurityProtocol securityProtocol = SecurityProtocol.SASL_PLAINTEXT;
+        configureMechanisms("PLAIN", Arrays.asList("PLAIN"));
+        server = NetworkTestUtils.createEchoServer(securityProtocol, saslServerConfigs);
+
+        // Create PLAINTEXT connection to manually authenticate after ApiVersionsRequest
+        String node = "1";
+        createClientConnection(SecurityProtocol.PLAINTEXT, node);
```

Comment:
@gwenshap looked into this and showed me why it doesn't work. Could we use a plain socket to send the api versions request to avoid the issue? It's a bit difficult to verify that we are doing the right thing with the current test.

For the second question, the current thinking is that we will do that after 0.10.0.0.